### DISCUSSION TITLE
Access modifiers on method parameters

### DISCUSSION THREAD

**Discussion by TizzyT566** [dotnet/csharplang#5416] [Language Ideas]
I was thinking it'd be nice for parameters to have access modified on them.
Restrictions would be that access modifiers can only be placed on all params or the params at the end similar to param arrays.
Access modifiers are grouped from most to least visible: internal>protected>private
public isn't used because it is the default.
The method's own access modifier cannot be more visible than the most visible parameter.
Parameters with access modifiers must have default values so the derived overloads have values to use.

```cs
// without access modifiers on params
public object Method(object p1) => Method(p1, new object(), new object());
private object Method(object p1, object p2, object p3)
{
   // code here
}

// with access modifiers on params
public object Method(object p1, private object p2 = new object(), private object p3 = new object())
{
   // code here
}
```

I also see no conflict with param arrays and default parameters, but I am not that knowledgeable under the hood so I maybe wrong.

```cs
public object Method(object p1, private object p2 = default, private params object[] p3)
{
   // code here
}
```
```cs
// because the code ultimately calls the main overload maybe constants aren't necessary?
public object M(object p1, internal object p2 = new object(), protected object p3 = null, private object p4 = new object()) { ... }

// would have the following equivalent overloads

public object M(object p1) => M(p1, new object(), null, new object());
internal object M(object p1, object p2) => M(p1, p2, null, new object());
protected object M(object p1, object p2, object p3) => M(p1, p2, p3, new object());
// main overload
private object M(object p1, object p2, object p3, object p4) { ... }
```

The closest I could find to this idea is:
https://github.com/dotnet/csharplang/discussions/804

but that seems more about automatically defining properties from parameters in the constructor.

**Comment by HaloFour** [comment]
So this is suggesting that the following syntax:

```cs
public void M(public int x, public int y, public int z) { }
```

would result in 4 different overloads of `M`?  That seems pretty confusing to me.

**Reply by TizzyT566** [reply]
No there is only 1 overload in this case. You can think of it as though the default access modifier for params is public.

**Reply by HaloFour** [reply]
That's even stranger and seems very niche as it would only work for overloads that happen to be a superset of the public overloads and have a different accessibility modifier.

**Comment by TizzyT566** [comment]
To explain a little bit further:

Within the same class for instance in the case of private you can use Method(...) with all of its params exposed, but outside of the class the params marked private will not be exposed.

In another example:
```cs
public object M(object p1, internal object p2) { ... }
```

would expose both params in the project. This is probably useful when making libraries as when using the method only the overloads not marked internal will be visible.

**Comment by ZacharyPatten** [comment]
Source generator instead of language feature? Seems appropriate here.

**Reply by TizzyT566** [reply]
I am not too familiar with source generators aside from just hearing of them but looking into it, it seems to be a compilation feature? does that mean that it will lack intellisense etc? source generators seem appropriate (from what I understand, which isn't much lol) for the library example I mentioned but what about within your own projects? Does this becoming a language feature not have a place?

**Reply by PathogenDavid** [reply]
> it seems to be a compilation feature?

You can think of them as compiler plugins that run just before the project is done compiling and can add code to it.

@davidwengier has a neat tool for tinkering with them if you want a quick way to get started: https://wengier.com/SourceGeneratorPlayground/

> does that mean that it will lack intellisense etc?

Intellisense works with Roslyn source generators.

<sub>(There is a quirk with Visual Studio in particular though where you have to restart in order to get changes reflected in Intellisense if you update a source generator.)</sub>

> but what about within your own projects?

This is probably more clear now if you looked at the tool I linked above, but since source generation happens before your compilation is complete, things added by a source generator to a project are usable within the same project.

(It isn't like IL weaving where the weaving takes place after the compilation is complete.)

> Does this becoming a language feature not have a place?

It still could, but language features take a *long* time to develop. If you use a source generator you can start doing this today.

**Reply by TizzyT566** [reply]
Thanks for the reply and material. I have looked into source generators a little more and looked at the playground. The playground was a great resource but having looked at the sample I have to ask. How does one go about implementing the idea I have presented via source generators? I would also love to hear from @davidwengier on the matter. Sorry if my suggestion is unclear and/or my lack of knowledge is paramount causing confusion.

What I am suggesting is turning multiple overloads which calls the main method into a single method. The other overloads are specified via the access modifiers. I had the idea when I was working on a method which needed to be called within the same class as well as outside. The method was made private and is recursive and had parameters I didn't want exposed to outside the class. So like usual I create another overload marked as public with only the parameters I want to expose outside and it calls the private version with default values for the "private" parameters.

I realize that this maybe niche but I feel like I've come across the situation enough times to at least suggest it. I also realize that language features are large endeavors and am not expecting this to be implemented just by simple suggestion. Mostly I am curious of any conflicts, alternatives, pros/cons etc. While an alternative has been brought forth like mentioned I can not conceive its implementation so further discussion in that regard would be great if possible. Thank you again.

**Reply by PathogenDavid** [reply]
No problem!

> I realize that this maybe niche but I feel like I've come across the situation enough times to at least suggest it.

Definitely nothing wrong with suggesting. Personally I do think it's too niche, but I do think it's an interesting idea since I've definitely written a lot of these boilerplate overloads before for similar reasons.

One issue with this feature in particular is there's some decisions that would be very hard for LDT to make. Some examples that come to mind:

* Is this feature still useful with non-const default argument values? (C# doesn't have these right now, but your examples use them. This would likely be considered a separate feature. Some people consider this to be an anti-feature that C# should not have.)
* Do you allow interleaving accessibilities? (IE: Can you make a method that's `public, private, internal, public`?)
* Should nullable parameters become non-nullable in the overloads? (For me it's common that the implementation method uses nullable for parameters only used in some overloads but I don't want them to be nullable in the overloads which have them.)
* Do default parameter values get exposed on all overloads? (Some people will want them to, some people only want them for the missing parameters for the call to the implementation.)
* Do the overloads perform parameter validation that the implementation method doesn't? (IE: The implementation might be tolerant of some parameters being null when the internal overloads should not.)

(These are rhetorical questions BTW, don't feel like you need to answer them. My point is there's no objectively correct answer to all of these. What's right for your codebase might not be correct for everyone.)

One of the benefits of source generators is that you can tailor them to your problem domain. You don't have to worry about what every C# developer ever needs, you just need to worry about what you/your coworkers/your clients/whoever need.

---------------

> How does one go about implementing the idea I have presented via source generators?

You're right that you will not be able to accomplish the syntax exactly as you specified above. I would suggest using attributes instead. Using your first example:

```csharp
public object M(object p1, internal object p2 = new object(), protected object p3 = null, private object p4 = new object())
    => throw new NotImplementedException();
```

would become something like:

```csharp
private object M(object p1, [Internal] object? p2 = null, [Protected] object? p3 = null, [Private] object? p4 = null)
    => throw new NotImplementedException();
```

(Note that I changed your default values to null since `new object()` is not possible as a default value. Even if this language proposal were accepted I would expect LDM to consider this a separate feature entirely. If you need this you could use the attributes to provide these values as expressions instead, IE: `[Internal(Default = "new object()")]`.)

Your source generator would then look for any method with parameters annotated with `[Private]`, `[Internal]`, etc. (Let's assume your source generator considers `p1` to be annotated as `[Public]` by default.)

It would be an error for any such methods to:

* Be defined on a non-partial class/struct. (`partial` is required for the source generator to add overloads.)
* Have an accessibility more accessible than any of the attributes used. (IE: A `public` method cannot have `[Private]` parameters. In your example your "public" method has private parameters, you need it to be private from the start since the source generator cannot change this.)
* (Technically optional, but I'd design it this way) Have any parameter which is more accessible than parameters before it in the parameter list. (IE: `public, public, internal, private, private`

For each method identified above:

* Generate a partial type definition for the method's containing type.
* For each `levelOfAccessibility` used on any parameter:
  * If `levelOfAccessibility` is the same as the original method, do nothing.
  * (Optional: Skip any method which already has a matching overload on the type. -- This might be helpful if you ever need to manually implement one of the overloads for some reason.)
  * Generate an overload for `levelOfAccessibility` without the parameters which are less accessible than `levelOfAccessibility` which calls the actual method.
    * (You might generate these methods with all non-public parameters changed to be non-nullable depending on what you think makes sense in your problem domain.)

So using the above example again, here's what the execution of that loop looks like:

* For each `levelOfAccessibility` in [`public`, `internal`, `protected`, `private`]:
  * `levelOfAccessibility = public`
    * Generate method: `public object M(object p1) => M(p1, null, null, null);`
  * `levelOfAccessibility = internal`
    * Generate method: `internal object M(object p1, object p2) => M(p1, p2, null, null);`
      * (Note that I dropped the nullable as I said above. You might also want to generate an argument check here if you don't want callers of the `internal` method to specify `null` for `p2`. I didn't do this for brevity.)
  * `levelOfAccessibility = protected`
    * Generate method: `protected object M(object p1, object p2, object p3) => M(p1, p2, p3, null);`
  * `levelOfAccessibility = private`
    * Generate nothing since the actual method is `private`.

So in the end you have a generated file that looks like this:

```csharp
partial class C
{
    public object M(object p1)
        => M(p1, null, null, null);

    internal object M(object p1, object p2)
        => M(p1, p2, null, null);

    protected object M(object p1, object p2, object p3)
        => M(p1, p2, p3, null);
}
```

(Also if you haven't stumbled upon it yet, there's a few more smaller source generator examples in the [Roslyn repo](https://github.com/dotnet/roslyn/blob/a4d1d8f0c7c2d723cae21198af5f0ec0723999ff/docs/features/source-generators.cookbook.md).)

**Reply by TizzyT566** [reply]
Thank you for your great reply and humoring this poor soul lol. I know I wasn't supposed to answer any of the points but I will just so its known what it is that's swirling around in my head.

> Is this feature still useful with non-const default argument values? (C# doesn't have these right now, but your examples use them. This would likely be considered a separate feature. Some people consider this to be an anti-feature that C# should not have.)

I originally followed the restrictions in place right now that requires the default values be constants but after thinking about I figured since they are just calling the "implementation method" with passed in values anyways we can let go of that restriction. Then I thought about how trouble some it maybe if say the values being lambdas or some other exotic input. I only left it there with the comment since I don't know what lies ahead if it were to be non-const-able.

> Do you allow interleaving accessibilities? (IE: Can you make a method that's public, private, internal, public?)

If I am understanding correctly then yes, It should accept overloads with implementations from the user over the generated overloads.
EDIT: After reading this again, I mean no. The access modifiers are grouped together from most accessible to least from left to right. Sorry if it was unclear in the original post but that was what I meant by
"Access modifiers are grouped from most to least visible: internal>protected>private".

> Should nullable parameters become non-nullable in the overloads? (For me it's common that the implementation method uses nullable for parameters only used in some overloads but I don't want them to be nullable in the overloads which have them.)

I hadn't thought about the nullable situation specifically but I did have in mind that the generated overloads have the exact param types as the implementation method, as in if the param is nullable then the generated one will also be nullable. I figured that would eliminate any complexities and preserve a sense of consistency seeing as you wouldn't actually see the generated overloads.

> Do default parameter values get exposed on all overloads? (Some people will want them to, some people only want them for the missing parameters for the call to the implementation.)

Yes the idea is that the generated overloads carry over the default values from the implementation method. I did originally make the default values not a requirement but some thinking later stumbled upon a scenario which would cause issue so I then specified that parameters (besides ones with the same access as the method itself since those params aren't going to have overloads for them and instead are part of the implementation method) would require default values. This I figure isn't an issue since one could manually create an overload and that would take precedence over a generated one like mentioned earlier.

> Do the overloads perform parameter validation that the implementation method doesn't? (IE: The implementation might be tolerant of some parameters being null when the internal overloads should not.)

The approach was a simplistic one so I hadn't thought about validations or any other things relating to extending the capabilities for generated overloads themselves. I only had the thought that the generated overloads just calls the implementation method and that would be it. So if the implementation method doesn't implement checking then the overloads which call it wont either, of course one is free to create their own overload manually which again takes precedence.

> You're right that you will not be able to accomplish the syntax exactly as you specified above. I would suggest using attributes instead.

Finally we get to the approach (yay) via source generators with parameter attributes. I originally did think about how this can be done alternatively and I figured Attributes would be a way but without knowledge of source generators I couldn't piece the two together. With that being said, I did figure that the syntax that I presented was going to be scrapped if implemented this way which is a shame since I think its much cleaner (imo). Having read your explanation it still seems that I am misunderstanding source generators to some degree and I will need to create some projects utilizing them to get familiar. I'd like to thank you again for your concise and whole hearted efforts.
No I did not stumble upon those example yet, will definitely take a look there as well.

**Reply by PathogenDavid** [reply]
No problem, glad I could help!

> I did figure that the syntax that I presented was going to be scrapped if implemented this way which is a shame since I think its much cleaner (imo).

Unfortunately it's either compromise on the syntax or wait until November 20XX for a new version of C# 😅

Good luck, have fun!
